keps/sig-api-machinery/20190324-declatative-field-manager.md

```diff
@@ -0,0 +1,207 @@
+---
+title: declarative-field-manager
+authors:
+  - "@kwiesmueller"
+owning-sig: sig-api-machinery
+reviewers:
+  - "@apelisse"
+  - "@lavalamp"
+approvers:
+  - "@lavalamp"
+creation-date: 2019-03-24
+last-updated: yyyy-mm-dd
+status: provisional
+---
+
+# declarative-field-manager
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Stories [optional]](#user-stories-optional)
+    - [Git Managed Configuration](#git-managed-configuration)
+    - [Kustomize](#kustomize)
+- [Design Details](#design-details)
+  - [Proposed API Change](#proposed-api-change)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+<!-- /toc -->
+
+## Summary
+
+Currently the identifier of the current actor (fieldManager) is set through an request option or the request user-agent,
+this KEP aims to provide a declarative way of setting the current fieldManager through a field in the object itself.
+
+## Motivation
+
+It is a common practice to use Kubernetes manifests as stored configuration of an application. Either to just introduce version control to the cluster's state, or to automate the process of updating objects.
+
+With the addition of Server Side Apply it could become necessary to change the way those manifests get applied, as the most benefit from this new feature requires every actor to define it's fieldManager. Doing this, currently is only possible through api options, the user-agent or as a kubectl option.
+
+To keep the way of interacting with manifests as easy as possible, this KEP suggests to allow defining the fieldManager as part of the manifest itself. Therefore no further action is required when applying a set of manifests and users can easily benefit from Server Side Apply without having to change their apply workflow.
```

Comment:
> Therefore no further action is required when applying a set of manifests and users can easily benefit from Server Side Apply without having to change their apply workflow.

It's not clear to me that a field manager option in the manifest accomplishes this.

`Apply` clients that accept and submit data from manifests [must pass a fieldManager query option](https://github.com/kubernetes/kubernetes/blob/release-1.17/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation.go#L117-L120). Adding a mutually exclusive fieldManager option in the manifest content would require those clients to introspect manifests to determine if they should set the query parameter. Requiring all apply clients to do that runs counter to `apply` being simple to implement client-side, and passing manifest data through as transparently as possible.

I'm also trying to think through the consequences of separating the `fieldManager` and `force` options. Changing `fieldManager` in a manifest would mean that any re-application of that manifest would conflict, and would depend on the applier doing a `--force` apply. If the dimensions of `apply` are "what" (manifest), "who" (fieldManager), and "how" (force or not), pairing fieldManager/force seems more natural than pairing manifest/fieldManager.

Reply:
> Adding a mutually exclusive fieldManager option in the manifest content would require those clients to introspect manifests to determine if they should set the query parameter.

That's a good point, and I agree that it is quite unsatisfying.

> Changing fieldManager in a manifest would mean that any re-application of that manifest would conflict, and would depend on the applier doing a --force apply. 

I see it from a different perspective. If the *source-of-truth* config committed in a github repository needs to be applied with a specific fieldManager (very likely, why wouldn't it be?), that knowledge has to be implicitly known and manually passed to kubectl every time someone runs `kubectl apply`, which is much more likely to cause conflict if done wrong (very likely). On the other hand, changing the content of the manifest to change the fieldManager would have to be reviewed/dry-run'd for failure and more likely to get caught.

Hopefully, we can soon live in a world were people don't have to invoke `kubectl apply` manually, but the tool still has to be configured somehow to know what fieldmanager to use, and that configuration will have to be loosely wired.

> pairing fieldManager/force seems more natural than pairing manifest/fieldManager.

IOW, I do believe that pairing manifest/fieldManger is *very* important and natural :-)